### PR TITLE
feat(connector): [BANKOFAMERICA] Implement 3DS flow for cards

### PR DESCRIPTION
**Pull request by deepanshu-iiitu**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Implement 3DS flow for cards

```
Legend: 
UB: User Browser
M: Merchant
HS: Hyperswitch
CS: Bankofamerica/Cybersource Connector
```

3DS FLOW DIAGRAM

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

https://github.com/juspay/hyperswitch/issues/3342

### QA Testing

Screenshot 2024-01-12 at 5 35 04 PM
Screenshot 2024-01-12 at 5 33 54 PM
Screenshot 2024-01-12 at 5 33 10 PM

Screenshot 2024-01-12 at 5 33 28 PM

- Test Non-3DS flows to confirm they are not affected
- Test 3DS transactions

All the possible 3DS test cases and test cards are available here: https://developer.cybersource.com/content/dam/docs/cybs/en-us/payer-authentication/developer/all/rest/payer-auth.pdf

3DS Payment Curl:
`curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: API_KEY_HERE' \
--data-raw '{
  "amount": 1404,
  "currency": "USD",
  "confirm": true,
  "capture_method": "automatic",
  "capture_on": "2022-09-10T10:11:12Z",
  "customer_id": "CustomerX",
  "email": "guest@example.com",
  "name": "John Doe",
  "phone": "999999999",
  "phone_country_code": "+1",
  "description": "Its my first payment request",
  "authentication_type": "three_ds",
  "return_url": "https://google.com",
  "payment_method": "card",
  "payment_method_type": "credit",
  "payment_method_data": {
    "card": {
       "card_number": "340000000001098",    
      "card_exp_month": "12",
      "card_exp_year": "30",
      "card_holder_name": "joseph",
      "card_cvc": "123"
    }
  },
  "billing": {
    "address": {
      "line1": "sdv",
      "line2": "Harrison Street",
      "line3": "Harrison Street",
      "city": "San Fransico",
      "state": "California",
      "zip": "46282",
      "country": "US",
      "first_name": "joseph",
      "last_name": "ewcjwd"
    },
    "phone": {
      "number": "8056594427",
      "country_code": "+97"
    }
  },
  "statement_descriptor_name": "joseph",
  "statement_descriptor_suffix": "JS",
  "metadata": {
    "count_tickets": "1",
    "transaction_number": 2233
  },
  "business_label": "food",
  "business_country": "US"
}'`

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Review by SamraatBansal — DISMISSED** [review]

**Review by srujanchikke — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/bankofamerica.rs [hunk 1]
```diff
@@ -316,13 +426,20 @@ impl ConnectorIntegration<api::Authorize, types::PaymentsAuthorizeData, types::P
...
         connectors: &settings::Connectors,
     ) -> CustomResult<String, errors::ConnectorError> {
-        Ok(format!(
-            "{}pts/v2/payments/",
-            api::ConnectorCommon::base_url(self, connectors)
-        ))
+        if req.is_three_ds() && req.request.is_card() {
+            Ok(format!(
+                "{}risk/v1/authentication-setups",
+                api::ConnectorCommon::base_url(self, connectors)
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/bankofamerica.rs]: Done

**Review by srujanchikke — DISMISSED** [review]

**Review by srujanchikke — DISMISSED** [review]

File: crates/router/src/connector/bankofamerica/transformers.rs [hunk 2]
```diff
@@ -367,6 +445,83 @@ pub struct Avs {
...
+{
+    type Error = error_stack::Report<errors::ConnectorError>;
+    fn try_from(
+        (item, ccard): (
+            &BankOfAmericaRouterData<&types::PaymentsCompleteAuthorizeRouterData>,
+            payments::Card,
+        ),
+    ) -> Result<Self, Self::Error> {
+        let email = item.router_data.request.get_email()?;
+        let bill_to = build_bill_to(item.router_data.get_billing()?, email)?;
```
Comment (ArjunKarthik) [inline crates/router/src/connector/bankofamerica/transformers.rs:462]: Do we need this to be added in dynamic fields? since this is for 3ds?

**Review by SamraatBansal — APPROVED** [review]

File: crates/router/src/core/payments.rs [hunk 3]
```diff
@@ -1495,6 +1495,22 @@ where
...
+                let should_continue = matches!(
+                    router_data.response,
+                    Ok(router_types::PaymentsResponseData::TransactionResponse {
+                        redirection_data: None,
+                        ..
+                    })
+                ) && router_data.status
+                    != common_enums::AttemptStatus::AuthenticationFailed;
+                (router_data, should_continue)
+            } else if connector.connector_name == router_types::Connector::Bankofamerica
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/core/payments.rs]: Done

File: crates/router/src/connector/bankofamerica/transformers.rs (hunk 2 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/bankofamerica/transformers.rs:462]: These fields are required in authorize call (even for 3ds). If we not pass these fields the connectors is throwing error.

Screenshot 2024-01-16 at 4 56 02 PM

File: crates/router/src/services/api.rs [hunk 4]
```diff
@@ -786,6 +786,15 @@ pub enum RedirectForm {
     Html {
         html_data: String,
     },
+    BankOfAmericaAuthSetup {
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/services/api.rs]: Done

**Review by ArjunKarthik — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by SamraatBansal — APPROVED** [review]

**Review by srujanchikke — APPROVED** [review]

**Pull request merged**